src/core/achievements.cpp

```diff
@@ -2059,6 +2061,170 @@ std::string Achievements::GetGameBadgePath(std::string_view badge_name)
+  else
+    http->CreateRequest(request.url, std::move(callback));
+
+  rc_api_destroy_request(&request);
+  http->WaitForAllRequests();
+
+  if (!request_success || response_data.empty())
+  {
+    Error::SetStringView(error, "Failed to fetch game info from RetroAchievements.");
+    return false;
+  }
+
+  // Parse response
+  rc_api_fetch_game_titles_response_t titles_response;
+  rc_api_server_response_t server_response;
+  server_response.body = reinterpret_cast<const char*>(response_data.data());
+  server_response.body_length = response_data.size();
+  server_response.http_status_code = 200;
+
+  const int parse_result = rc_api_process_fetch_game_titles_server_response(&titles_response, &server_response);
+  if (parse_result != RC_OK)
+  {
+    const std::string_view response_preview(server_response.body,
+                                            std::min<size_t>(server_response.body_length, 500));
+    ERROR_LOG("Failed to parse game titles response ({}): {}", parse_result, response_preview);
+    if (titles_response.response.error_message)
+      Error::SetStringFmt(error, "RetroAchievements error: {}", titles_response.response.error_message);
+    else
+      Error::SetStringFmt(error, "Failed to parse API response (code {})", parse_result);
+    rc_api_destroy_fetch_game_titles_response(&titles_response);
+    return false;
+  }
+
+  ScopedGuard response_guard([&titles_response]() { rc_api_destroy_fetch_game_titles_response(&titles_response); });
+
+  if (titles_response.num_entries == 0)
+  {
+    progress->SetStatusText("No icon information found.");
+    return true;
+  }
+
+  // Download each icon
+  progress->SetProgressRange(titles_response.num_entries);
+  u32 downloaded = 0;
+
+  for (u32 i = 0; i < titles_response.num_entries; i++)
+  {
+    if (progress->IsCancelled())
+      break;
+
+    const rc_api_game_title_entry_t& entry = titles_response.entries[i];
+    progress->SetProgressValue(i);
+    progress->SetStatusText(fmt::format("Downloading icon for {}...", entry.title).c_str());
+
+    if (!entry.image_name || entry.image_name[0] == '\0')
+      continue;
+
+    const std::string local_path = GetLocalImagePath(entry.image_name, RC_IMAGE_TYPE_GAME);
+    if (FileSystem::FileExists(local_path.c_str()))
+    {
+      GameList::UpdateAchievementBadgeName(entry.id, entry.image_name);
+      continue;
+    }
+
+    const std::string url = GetImageURL(entry.image_name, RC_IMAGE_TYPE_GAME);
+    if (url.empty())
+      continue;
+
+    std::vector<u8> image_data;
+    bool download_success = false;
+
+    http->CreateRequest(url, [&image_data, &download_success](s32 status_code, const Error&, const std::string&,
+                                                              HTTPDownloader::Request::Data data) {
+      if (status_code == HTTPDownloader::HTTP_STATUS_OK)
+      {
+        image_data = std::move(data);
+        download_success = true;
+      }
+    });
+    http->WaitForAllRequests();
```

Comment:
This will block for each request, which would be slow on high-latency connections. Would be better to create all requests in parallel. Can do this later if you're not sure what I mean.